p11-kit/export-object.c

```diff
@@ -59,85 +67,333 @@ int
 p11_kit_export_object (int argc,
 		       char *argv[]);
 
+#ifdef WITH_ASN1
+
 static void
-export_attribute (P11KitIter *iter,
-		  CK_ATTRIBUTE_TYPE type,
-		  const char *label)
+prepend_leading_zero (CK_ATTRIBUTE *attr)
 {
-	p11_buffer buf;
-	CK_ATTRIBUTE attr = { type, NULL_PTR, 0 };
+	if (*((unsigned char *)attr->pValue) & 0x80) {
+		unsigned char *padded;
+
+		return_if_fail (attr->ulValueLen < ULONG_MAX);
+		padded = malloc (attr->ulValueLen + 1);
+		return_if_fail (padded);
+		memcpy (padded + 1, attr->pValue, attr->ulValueLen);
+		*padded = 0x00;
+		free (attr->pValue);
+		attr->pValue = padded;
+		attr->ulValueLen++;
+	}
+}
 
-	if (!p11_buffer_init (&buf, 0)) {
-		p11_message (_("failed to initialize buffer"));
-		return;
+static bool
+export_pubkey_rsa (P11KitIter *iter, p11_buffer *buf)
+{
+	CK_ATTRIBUTE template[] = {
+		{ CKA_MODULUS, },
+		{ CKA_PUBLIC_EXPONENT, },
+	};
+	CK_ATTRIBUTE *attrs = NULL;
+	CK_ATTRIBUTE *modulus, *public_exponent;
+	asn1_node asn = NULL;
+	p11_dict *defs = NULL;
+	int result;
+	unsigned char *spk = NULL, *der = NULL;
+	const unsigned char null[] = { 0x05, 0x00 };
+	size_t n_spk, n_der;
+	bool ok = false;
+
+	attrs = p11_attrs_buildn (NULL, template, 2);
+	return_val_if_fail (attrs, false);
+
+	if (p11_kit_iter_load_attributes (iter, attrs, p11_attrs_count (attrs)) != CKR_OK) {
+		p11_message (_("failed to retrieve attributes"));
+		goto cleanup;
 	}
 
-	if (p11_kit_iter_get_attributes (iter, &attr, 1) != CKR_OK) {
-		p11_message (_("failed to retrieve attribute length of an object"));
+	modulus = p11_attrs_find_valid (attrs, CKA_MODULUS);
+	public_exponent = p11_attrs_find_valid (attrs, CKA_PUBLIC_EXPONENT);
+
+	if (!modulus || !public_exponent) {
+		p11_message (_("failed to retrieve attributes"));
 		goto cleanup;
 	}
 
-	attr.pValue = malloc (attr.ulValueLen);
-	if (attr.pValue == NULL) {
-		p11_message (_("failed to allocate memory"));
+	defs = p11_asn1_defs_load ();
+	return_val_if_fail (defs, false);
+
+	asn = p11_asn1_create (defs, "PKIX1.RSAPublicKey");
+	return_val_if_fail (asn, false);
+
+	prepend_leading_zero (modulus);
+	result = asn1_write_value (asn, "modulus", modulus->pValue, modulus->ulValueLen);
+	if (result != ASN1_SUCCESS) {
+		p11_message (_("unable to write value"));
+		goto cleanup;
+	}
+	prepend_leading_zero (public_exponent);
+	result = asn1_write_value (asn, "publicExponent", public_exponent->pValue, public_exponent->ulValueLen);
+	if (result != ASN1_SUCCESS) {
+		p11_message (_("unable to write value"));
 		goto cleanup;
 	}
 
-	if (p11_kit_iter_get_attributes (iter, &attr, 1) != CKR_OK) {
-		p11_message (_("failed to retrieve attribute of an object"));
+	spk = p11_asn1_encode (asn, &n_spk);
+	if (!spk) {
+		p11_message (_("unable to encode %s"), "subjectPublicKey");
+		goto cleanup;
+	}
+	p11_asn1_free (asn);
+
+	asn = p11_asn1_create (defs, "PKIX1.SubjectPublicKeyInfo");
+	return_val_if_fail (asn, false);
```

Comment:
Wouldn't this leak `spk`?

Reply:
It leaks, but I would say it's ok because this is not a library code. We use this macro in such occasions (untestable error in commands) already.

Reply:
I would prefer to still `goto cleanup` in such cases.

Reply:
Note that those macros are designed as a counterpart `g_return_if_*`, which are used like an assert: https://developer-old.gnome.org/glib/stable/glib-Warnings-and-Assertions.html

As they will cause the program abort if `P11_KIT_STRICT` is set (similar to `G_DEBUG=fatal_*` is set), we should already have a sufficient test coverage even if we dont't do the unrealistic cleanup. I believe other low-level commands (e.g., coreutils) do similar, e.g., they use `xmalloc` which simply aborts the program when memory allocation fails.

Reply:
As you say, they are used like an assert, which guards against programming errors, not normal runtime errors like memory allocation failure and such. From `g_return_if_*` _"If expr evaluates to [FALSE](https://developer-old.gnome.org/glib/stable/glib-Standard-Macros.html#FALSE:CAPS), the current function should be considered to have undefined behaviour (a programmer error)"_. If `asn = p11_asn1_create (defs, "PKIX1.SubjectPublicKeyInfo");` fails, thats not a programming error. I think we should cleanup in such cases.

Reply:
I argue against that. The failure in `asn = p11_asn1_create (defs, "PKIX1.SubjectPublicKeyInfo");`  really shouldn't happen and we can't do anything on that. That is a kind of undefined behavior in this case. You can see the exact same pattern in the `trust extract` command: https://github.com/p11-glue/p11-kit/blob/e8872bfdc4b12735bbc0433a59fc0a7ddfd68bbb/trust/extract-openssl.c#L299

Reply:
In the end I refactored the code by splitting the function into smaller ones, so we don't need `goto` for some cases, and proper error handling even in unrealistic cases. In the future maybe we should consider using automatic cleanup as suggested on #577.